### PR TITLE
add --uri-substitution option

### PR DESCRIPTION
**Pull request by jakemac53**

### CODE REVIEW

File: dart/tools/dev_server/bin/server.dart [hunk 1]
```diff
@@ -61,6 +67,20 @@ Future main(List<String> args) async {
...
+/// Map of uri substitutions to be applied in [_applyUriSubstitutions]
+final _uriSubstitutions = <String, String>{};
+
+/// [Handler] that applies [_uriSubstitutions].
+Handler _applyUriSubstitutions(Handler innerHandler) => (Request request) {
+      var newPath = request.requestedUri.path;
+      _uriSubstitutions.forEach((from, to) {
+        newPath = newPath.replaceFirst(from, to);
+      });
+      if (newPath == request.requestedUri.path) return innerHandler(request);
```
Comment (kevmoo) [inline dart/tools/dev_server/bin/server.dart]: do a reasign to `request` here then you only need one call to `innerHandler`

File: dart/tools/dev_server/bin/server.dart [hunk 2]
```diff
@@ -85,19 +105,22 @@ Handler _reroutePackagesPaths(Handler innerHandler) => (Request request) {
...
-          context: request.context);
+      var newRequest = _changeRequestPath(
+          request, request.requestedUri.replace(path: filePath));
       return innerHandler(newRequest);
     };
 
+/// Takes a [Request] and updates its uri to [newUri].
+/// TODO: We should be using `original.change` but its overly restrictive.
+Request _changeRequestPath(Request original, Uri newUri) =>
+    new Request(original.method, newUri,
```
Comment (kevmoo) [inline dart/tools/dev_server/bin/server.dart]: noice

File: dart/tools/dev_server/bin/server.dart (hunk 1 above)
Comment (jakemac53) [inline dart/tools/dev_server/bin/server.dart]: done

**Comment by jakemac53** [comment]
rebased and updated

File: dart/tools/dev_server/bin/server.dart [hunk 3]
```diff
@@ -61,6 +67,20 @@ Future main(List<String> args) async {
   }
 }
 
+/// Map of uri substitutions to be applied in [_applyUriSubstitutions]
+final _uriSubstitutions = <String, String>{};
+
+/// [Handler] that applies [_uriSubstitutions].
```
Comment (natebosch) [inline dart/tools/dev_server/bin/server.dart]: [nit] this isn't a [Handler], its a [Middleware]

not sure this doc adds much new info though

File: dart/tools/dev_server/bin/server.dart [hunk 4]
```diff
@@ -62,6 +68,22 @@ Future main(List<String> args) async {
...
+Handler _applyUriSubstitutions(Handler innerHandler) => (Request request) {
+      var newPath = request.requestedUri.path;
+      _uriSubstitutions.forEach((from, to) {
+        newPath = newPath.replaceFirst(from, to);
+      });
+      if (newPath == request.requestedUri.path) {
```
Comment (kevmoo) [inline dart/tools/dev_server/bin/server.dart]: `==` or `!=`?

File: dart/tools/dev_server/bin/server.dart (hunk 3 above)
Comment (kevmoo) [inline dart/tools/dev_server/bin/server.dart]: It'd be good if we didn't crash in checked mode. 😄

Comment (jakemac53) [inline dart/tools/dev_server/bin/server.dart]: lol good catch, I keep getting confused about the whole middleware/handler thing

Comment (kevmoo) [inline dart/tools/dev_server/bin/server.dart]: Or lie about the impl

File: dart/tools/dev_server/bin/server.dart (hunk 4 above)
Comment (jakemac53) [inline dart/tools/dev_server/bin/server.dart]: DOH!

**Comment by jakemac53** [comment]
Ok tested with https://github.com/dart-lang/bazel/pull/35 on the codelab app, everything works!

**Pull request merged**